galsim/detectors.py

```diff
@@ -148,5 +148,133 @@ def addReciprocityFailure(self, exp_time, alpha, base_flux):
+                                   original pixel values are retained at the edges. If
+                                   edge_treatment is not 'crop', then this is ignored.
+    @param kernel_nonnegativity    Specify whether the kernel should have only non-negative
+                                   entries.  [default: True]
+    @param kernel_normalization    Specify whether to check and enforce correct normalization for
+                                   the kernel.  [default: True]
+
+    @returns None
+    """
+
+    # IPC kernel has to be a 3x3 Image instance
+    if not isinstance(IPC_kernel,galsim.Image):
+        raise ValueError("IPC_kernel must be an Image instance .")
+    ipc_kernel = IPC_kernel.array
+    if not ipc_kernel.shape==(3,3):
+        raise ValueError("IPC kernel must be an Image instance of size 3x3.")
+
+    # Check for non-negativity of the kernel
+    if kernel_nonnegativity is True:
+        if (ipc_kernel<0).any() is True:
+            raise ValueError("IPC kernel must not contain negative entries")
+
+    # Check and enforce correct normalization for the kernel
+    if kernel_normalization is True:
+        if abs(ipc_kernel.sum() - 1.0) > 10.*numpy.finfo(ipc_kernel.dtype.type).eps:
+            import warnings
+            warnings.warn("The entries in the IPC kernel did not sum to 1. Scaling the kernel to "\
+                +"ensure correct normalization.")
+            IPC_kernel = IPC_kernel/ipc_kernel.sum()
+
+    # edge_treatment can be 'extend', 'wrap' or 'crop'
+    if edge_treatment=='crop':
+        # Simply re-label the array of the Image instance
+        pad_array = self.array
+    elif edge_treatment=='extend':
+        # Copy the array of the Image instance and pad with zeros
+        pad_array = numpy.zeros((self.array.shape[0]+2,self.array.shape[1]+2))
+        pad_array[1:-1,1:-1] = self.array
+    elif edge_treatment=='wrap':
+        # Copy the array of the Image instance and pad with zeros initially
+        pad_array = numpy.zeros((self.array.shape[0]+2,self.array.shape[1]+2))
+        pad_array[1:-1,1:-1] = self.array
+        # and wrap around the edges
+        pad_array[0,:] = pad_array[-2,:]
+        pad_array[-1,:] = pad_array[1,:]
+        pad_array[:,0] = pad_array[:,-2]
+        pad_array[:,-1] = pad_array[:,1]
+    else:
+        raise ValueError("edge_treatment has to be one of 'extend', 'wrap' or 'crop'. ")
+
+    # Generating different segments of the padded array
+    center = pad_array[1:-1,1:-1]
+    top = pad_array[2:,1:-1]
+    bottom = pad_array[:-2,1:-1]
+    left = pad_array[1:-1,:-2]
+    right = pad_array[1:-1,2:]
+    topleft = pad_array[2:,:-2]
+    bottomright = pad_array[:-2,2:]
+    topright = pad_array[2:,2:]
+    bottomleft = pad_array[:-2,:-2]
+
+    # Ensure that the choice of origin does not matter
+    x0 = IPC_kernel.bounds.xmin # 1 by default
+    y0 = IPC_kernel.bounds.ymin # 1 by default
```

Comment:
The code looks clearner with Image elements being used instead of the array elements since the bottom left pixel on the kernel (as an Image) is multiplied with the bottom left part of the Image and so on. If the array elements are used, then it can potentially be confusing later on.